File: symfony-app/src/Service/ForbiddenWordsFilterService.php
<?php

namespace App\Service;

class ForbiddenWordsFilterService
{
    /**
     * @param string[] $forbiddenWords
     */
    public function __construct(private array $forbiddenWords = [])
    {
    }

    /**
     * @return string[] matched forbidden words
     */
    public function findForbiddenWords(string $text): array
    {
        if ($text === '' || $this->forbiddenWords === []) {
            return [];
        }

        $cleaned = $this->cleanText($text);
        $matches = [];
        foreach ($this->forbiddenWords as $word) {
            $word = trim((string) $word);
            if ($word === '') {
                continue;
            }
            if ($this->matchesWord($cleaned, $word)) {
                $matches[] = $word;
            }
        }

        return array_values(array_unique($matches));
    }

    public function hasBadWord(string $text): bool
    {
        return $this->findForbiddenWords($text) !== [];
    }

    public function cleanText(string $text): string
    {
        $text = trim($text);
        $text = preg_replace('/\s+/u', ' ', $text) ?? $text;
        $text = preg_replace('/[^\p{L}\p{N}\s]/u', '', $text) ?? $text;
        return mb_strtolower($text);
    }

    private function matchesWord(string $cleanedText, string $word): bool
    {
        $normalizedWord = $this->normalizeLeetToken($word);
        if ($normalizedWord === '') {
            return false;
        }

        $tokens = preg_split('/\s+/u', $cleanedText, -1, PREG_SPLIT_NO_EMPTY) ?: [];
        foreach ($tokens as $token) {
            if ($this->normalizeLeetToken($token) === $normalizedWord) {
                return true;
            }
        }

        return false;
    }

    private function normalizeLeetToken(string $value): string
    {
        $value = mb_strtolower($value);
        $value = strtr($value, [
            '0' => 'o',
            '1' => 'i',
            '2' => 'e',
            '3' => 'e',
            '4' => 'a',
            '5' => 's',
            '6' => 'g',
            '7' => 't',
            '8' => 'b',
            '9' => 'g',
        ]);

        return preg_replace('/[^\p{L}\p{N}]/u', '', $value) ?? '';
    }
}

File: javafx-app/src/main/java/com/santea/service/ForbiddenWordsFilterService.java
package com.santea.service;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Filtre mots interdits (fallback modération) — aligné sur symfony-app ForbiddenWordsFilterService.
 */
public class ForbiddenWordsFilterService {
    private final List<String> forbiddenWords;

    public ForbiddenWordsFilterService(List<String> forbiddenWords) {
        this.forbiddenWords = forbiddenWords == null ? List.of() : List.copyOf(forbiddenWords);
    }

    public List<String> findForbiddenWords(String text) {
        if (text == null || text.isEmpty() || forbiddenWords.isEmpty()) {
            return List.of();
        }

        String cleaned = cleanText(text);
        Set<String> matches = new LinkedHashSet<>();
        for (String word : forbiddenWords) {
            String w = word == null ? "" : word.trim();
            if (w.isEmpty()) {
                continue;
            }
            if (matchesWord(cleaned, w)) {
                matches.add(w);
            }
        }
        return new ArrayList<>(matches);
    }

    public boolean hasBadWord(String text) {
        return !findForbiddenWords(text).isEmpty();
    }

    public String cleanText(String text) {
        if (text == null) {
            return "";
        }
        String t = text.trim().replaceAll("\\s+", " ");
        t = t.replaceAll("[^\\p{L}\\p{N}\\s]", "");
        return t.toLowerCase(Locale.ROOT);
    }

    private boolean matchesWord(String cleanedText, String word) {
        String normalizedWord = normalizeLeetToken(word);
        if (normalizedWord.isEmpty()) {
            return false;
        }

        String[] tokens = cleanedText.split("\\s+");
        for (String token : tokens) {
            if (token.isEmpty()) {
                continue;
            }
            if (normalizeLeetToken(token).equals(normalizedWord)) {
                return true;
            }
        }
        return false;
    }

    private String normalizeLeetToken(String value) {
        if (value == null) {
            return "";
        }
        String normalized = value.toLowerCase(Locale.ROOT)
                .replace('0', 'o')
                .replace('1', 'i')
                .replace('2', 'e')
                .replace('3', 'e')
                .replace('4', 'a')
                .replace('5', 's')
                .replace('6', 'g')
                .replace('7', 't')
                .replace('8', 'b')
                .replace('9', 'g');
        return normalized.replaceAll("[^\\p{L}\\p{N}]", "");
    }
}
